File: content/snippets/go/interview_lcci/16/02_words_frequency.go
package main

import "fmt"

type WordsFrequency struct {
	counter map[string]int
}

func Constructor(book []string) WordsFrequency {
	counter := map[string]int{}
	for _, word := range book {
		counter[word]++
	}
	return WordsFrequency{counter: counter}
}

func (w *WordsFrequency) Get(word string) int {
	return w.counter[word]
}

func main() {
	wf := Constructor([]string{"i", "have", "an", "apple", "he", "have", "a", "pen"})
	fmt.Println(wf.Get("have"))
	fmt.Println(wf.Get("you"))
}

File: content/snippets/cpp/interview_lcci/16/02_words_frequency.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>
using namespace std;

class WordsFrequency {
public:
    explicit WordsFrequency(const vector<string>& book) {
        for (const string& word : book) {
            ++counter[word];
        }
    }

    int get(const string& word) const {
        auto it = counter.find(word);
        return it == counter.end() ? 0 : it->second;
    }

private:
    unordered_map<string, int> counter;
};

int main() {
    vector<string> book = {"i", "have", "an", "apple", "he", "have", "a", "pen"};
    WordsFrequency wf(book);
    cout << wf.get("have") << endl;
    cout << wf.get("you") << endl;
    return 0;
}
